src/main/java/twigkit/frame/CachedImageIOService.java

```diff
@@ -70,104 +77,184 @@ public void setOfflinePath(String offlinePath) {
         }
     }
 
+    public void setCacheName(String cacheName) {
+        if (cacheName != null & !cacheName.equals("")) {
+            this.cacheName = cacheName;
+            createCacheManager();
+            cache = getOrCreateCache();
+        }
+    }
+
+    public void createCacheManager() {
+        URL url = getClass().getResource("/cache.xml");
+        this.cacheManager = CacheManager.newInstance(url);
+    }
+
+    public Ehcache getOrCreateCache() {
+        if (!cacheManager.cacheExists(cacheName)) {
+            logger.debug("Creating cache from defaults!");
+            cacheManager.addCache(cacheName);
+            Cache cache = (Cache) cacheManager.getEhcache(cacheName);
+            // TODO Seems to have no effect?
+            cache.getCacheConfiguration().setCopyOnRead(true);
+            cache.getCacheConfiguration().setCopyOnWrite(true);
+            if (logger.isTraceEnabled()) {
+//            if (!cache.isStatisticsEnabled()) {
+//                cache.setStatisticsEnabled(true);
+//            }
+                logger.trace("Cache [" + cacheName + "] TimeToIdleSeconds: " + cache.getCacheConfiguration().getTimeToIdleSeconds());
+            }
+            CacheEventListener myListener = new CacheEventListenerImpl();
+            cache.getCacheEventNotificationService().registerListener(myListener);
+            return cache;
+        } else {
+            return cacheManager.getEhcache(cacheName);
+        }
+    }
+
     @Override
     public Image fromURL(final URL url) throws IOException {
         return fromURL(url, true);
     }
 
-	public Image fromURL(final URL url, boolean writeToCache) throws IOException {
-		if (repository != null && repository.exists()) {
-			if (logger.isTraceEnabled()) {
-				logger.trace("Getting Image from cache [" + repository.getAbsolutePath() + "]");
-			}
-
-			File file = getFileFromURL(url, 0, 0);
-
-			if (file.exists()) {
-				if (logger.isTraceEnabled()) {
-					logger.trace("Getting Image [" + file.getName() + "] from cache");
-				}
-				BufferedImage buf = ImageIO.read(file);
-				Image image = new Image(buf);
-				image.setUrl(url);
-				return image;
-			} else {
-				Image image = super.fromURL(url);
-				image.setUrl(url);
-
-                if (writeToCache) {
-                    ImageIO.write(image.getBufferedImage(), Image.ContentType.PNG.getSuffix(), file);
-
-                    if (logger.isTraceEnabled()) {
-                        logger.trace("Wrote Image (original) [" + file.getName() + ", " + image.getWidth() + "px by " + image.getHeight() + "px] to cache");
-                    }
+    public Image fromURL(final URL url, boolean writeToCache) throws IOException {
+        Boolean isMissing = false;
+        String key = getKeyFromURLBySize(url, 0, 0);
+        if (writeToCache && repository != null && repository.exists() && cache != null && cache.isKeyInCache(key)) {
```

Comment:
Why are we checking `writeToCache` when reading from the cache? Am I missing something?